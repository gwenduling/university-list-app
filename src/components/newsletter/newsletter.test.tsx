import NewsLetter from "./newsletter";

import { shallow } from "enzyme";
import { DebounceInput } from "react-debounce-input";
import { NotificationManager } from "react-notifications";
import flushPromises from "flush-promises";

describe("Newsletter Component", () => {
  let wrapper: any;
  beforeEach(() => {
    wrapper = shallow(<NewsLetter />);
    wrapper.instance().textInput = {
      current: {
        focus: () => {},
      },
    };

    jest.spyOn(NotificationManager, "error");
    jest.spyOn(NotificationManager, "success");
  });

  afterEach(() => {
    jest.restoreAllMocks();
  });

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should updates the state on input change", () => {
    const value = "email@email.com";

    wrapper.find(DebounceInput).simulate("change", { target: { value } });
    expect(wrapper.state("email")).toEqual(value);
  });

  it("should trigger button on click", () => {
    jest.spyOn(wrapper.instance(), "subscribeToNewsletter");
    wrapper.find("button").simulate("click");
    expect(wrapper.instance().subscribeToNewsletter).toHaveBeenCalledTimes(1);
  });

  it("should notify error when email is empty string", () => {
    wrapper.instance().subscribeToNewsletter();

    expect(NotificationManager.error).toHaveBeenCalledWith(
      "Please enter valid email address."
    );
    expect(NotificationManager.error).toHaveBeenCalledTimes(1);
    expect(NotificationManager.success).toHaveBeenCalledTimes(0);
  });

  it("should notify success on success", async () => {
    // @ts-ignore
    global.fetch = jest.fn(async () =>
      Promise.resolve({
        ok: true,
        json: async () => ({ status: 200 }),
      })
    );

    wrapper.setState({ email: "test@test.com" });
    wrapper.instance().subscribeToNewsletter();
    await flushPromises();

    expect(NotificationManager.success).toHaveBeenCalled();
    expect(NotificationManager.error).not.toHaveBeenCalled();
  });

  it("should notify on duplicate email", async () => {
    // @ts-ignore
    global.fetch = jest.fn(async () =>
      Promise.resolve({
        json: () => {},
        status: 409,
      })
    );

    wrapper.setState({ email: "test@test.com" });
    wrapper.instance().subscribeToNewsletter();
    await flushPromises();

    expect(NotificationManager.success).not.toHaveBeenCalled();
    expect(NotificationManager.error).toHaveBeenCalledWith(
      "Email address already exists! Please use a different email address."
    );
  });

  it("should notify on error response from server", async () => {
    // @ts-ignore
    global.fetch = jest.fn(async () =>
      Promise.resolve({
        json: () => {},
        status: 500,
      })
    );

    wrapper.setState({ email: "test@test.com" });
    wrapper.instance().subscribeToNewsletter();
    await flushPromises();

    expect(NotificationManager.success).not.toHaveBeenCalled();
    expect(NotificationManager.error).toHaveBeenCalledWith(
      "Something went wrong! Please try again in a few minutes."
    );
  });

  it("should notify on promise error", async () => {
    // @ts-ignore
    global.fetch = jest.fn(() =>
      Promise.reject({
        message: "Something went wrong! Please try again in a few minutes.",
      })
    );

    wrapper.setState({ email: "test@test.com" });
    wrapper.instance().subscribeToNewsletter();
    await flushPromises();

    expect(NotificationManager.success).not.toHaveBeenCalled();
    expect(NotificationManager.error).toHaveBeenCalledWith(
      "Something went wrong! Please try again in a few minutes."
    );
  });
});
