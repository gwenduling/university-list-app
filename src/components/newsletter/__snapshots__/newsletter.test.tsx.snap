// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Newsletter Component should match the snapshot 1`] = `
<div
  className="newsletter-component _padding"
>
  <NotificationContainer
    enterTimeout={400}
    leaveTimeout={400}
  />
  <p
    className="title"
  >
    Don't miss an update!
  </p>
  <p
    className="content"
  >
    Join our newsletter.
  </p>
  <div
    className="form-item"
  >
    <DebounceInput
      debounceTimeout={200}
      element="input"
      forceNotifyByEnter={true}
      forceNotifyOnBlur={true}
      inputRef={
        Object {
          "current": null,
        }
      }
      minLength={0}
      onChange={[Function]}
      placeholder="Email Address"
      type="email"
      value=""
    />
    <button
      onClick={[Function]}
      type="submit"
    >
      Subscribe
    </button>
  </div>
</div>
`;
